pages/signup.tsx

```diff
@@ -0,0 +1,136 @@
+/**
+ * Page for Signing up the User
+ */
+// React
+import { useState, useContext } from 'react';
+import { useRouter } from 'next/router';
+
+// Next
+import type {
+  GetServerSideProps,
+  InferGetServerSidePropsType,
+  GetServerSidePropsContext,
+} from 'next';
+
+// Auth
+import UserContext from '../context/user/UserContext';
+
+// 3rd Party
+import { Button, Input, InputLabel, FormControl } from '@mui/material';
+
+// Components
+import ProtectedRoute from '../components/ProtectedRoute';
+import PopUp from '../components/PopUp';
+import AuthErrors from '../components/AuthErrors';
+
+const Signup = (
+  props: InferGetServerSidePropsType<typeof getServerSideProps>
+) => {
+  const router = useRouter();
+  const { isAuthenticated, signup } = useContext(UserContext);
+
+  const [data, setData] = useState({
+    email: '',
+    password: '',
+  });
+  const [open, setOpen] = useState(false);
+  const [title, setTitle] = useState('');
+  const [errorMessage, setErrorMessage] = useState('');
+
+  const handleOpen = () => {
+    setOpen(true);
+  };
+
+  const handleClose = () => {
+    setOpen(false);
+  };
+
+  const handleSignup = async (e: any) => {
+    e.preventDefault();
+    const res = await signup(data);
+    const authResult = AuthErrors(res);
+    if (authResult.authFailure) {
+      setTitle(authResult.title);
+      setErrorMessage(authResult.errorMessage);
+      handleOpen();
+    } else {
+      if (isAuthenticated) {
+        router.push(authResult.dest);
+      }
+    }
+  };
+
+  return (
+    <div
+      style={{
+        display: 'flex',
+        flexDirection: 'column',
+        justifyContent: 'space-around',
+        alignItems: 'center',
+      }}
+    >
+      <h1 className="text-center my-3 ">Signup</h1>
+      <form style={{ height: '100%' }} onSubmit={handleSignup}>
+        <div>
+          <FormControl>
+            <InputLabel>Email address</InputLabel>
+            <Input
+              type="email"
+              placeholder="Enter email"
+              required
+              onChange={(e: any) =>
+                setData({
+                  ...data,
+                  email: e.target.value,
+                })
```

Comment:
```suggestion
                setData((peviousData) => ({
                  ...previousData,
                  email: e.target.value,
                }))
```
When referencing a previous state in set state it is better practice to use the callback functionality of set state which receives the previous data as a parameter. This is to prevent race conditions making the react code more reliable in general. Although you might not run into any race conditions here it is still good practice to reformat your set states like this when referencing previous data. Here's a [link to read more](https://reactjs.org/docs/react-component.html#setstate:~:text=Subsequent%20calls%20will%20override%20values%20from%20previous%20calls%20in%20the%20same%20cycle%2C%20so%20the%20quantity%20will%20only%20be%20incremented%20once.%20If%20the%20next%20state%20depends%20on%20the%20current%20state%2C%20we%20recommend%20using%20the%20updater%20function%20form%2C%20instead%3A).